### PR TITLE
[NW Reconfig] add header to network  requests with epoch

### PR DESCRIPTION
**Pull request by laura-makdah**

### CODE REVIEW

File: narwhal/primary/src/primary.rs [hunk 1]
```diff
@@ -268,6 +271,10 @@ impl Primary {
                 inbound_network_metrics,
             )))
             .layer(CallbackLayer::new(FailpointsMakeCallbackHandler::new()))
+            .layer(SetRequestHeaderLayer::overriding(
```
Comment (akichidis) [inline narwhal/primary/src/primary.rs:274]: Should this be on the `service` or on the `outbound_layer` instead? We want to add the `epoch` header on the outbound requests right?

File: narwhal/worker/src/worker.rs [hunk 2]
```diff
@@ -177,6 +180,10 @@ impl Worker {
                 inbound_network_metrics,
             )))
             .layer(CallbackLayer::new(FailpointsMakeCallbackHandler::new()))
+            .layer(SetRequestHeaderLayer::overriding(
+                "epoch".parse().unwrap(),
```
[the remark this answers is not in the document]
Comment (laura-makdah) [reply narwhal/worker/src/worker.rs:184]: Yes, I agree. That is already in my next commit.

File: narwhal/primary/src/primary.rs (hunk 1 above)
Comment (laura-makdah) [reply narwhal/primary/src/primary.rs:274]: good call, moved to outbound

File: narwhal/network/src/epoch_filter.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        let epoch = request
+            .headers()
+            .get(EPOCH_HEADER_KEY)
+            .ok_or_else(|| StatusCode::InternalServerError.into_response())?;
+
+        if self.allowed_epoch == *epoch {
+            Ok(())
+        } else {
+            Err(StatusCode::VersionNotSupported.into_response())
+        }
```
[the remark this answers is not in the document]
Comment (bmwill) [reply narwhal/network/src/epoch_filter.rs:34]: In addition to that you could use https://github.com/bmwill/anemo/blob/main/crates/anemo/src/rpc/mod.rs#LL35C13-L35C13 in order to provide a message detailing the error for the client. Such a message could share why the request was denied, what version they sent vs what was expected.

File: narwhal/primary/src/primary.rs [hunk 4]
```diff
@@ -280,6 +291,10 @@ impl Primary {
                 outbound_network_metrics,
             )))
             .layer(CallbackLayer::new(FailpointsMakeCallbackHandler::new()))
+            .layer(SetRequestHeaderLayer::overriding(
+                EPOCH_HEADER_KEY.parse().unwrap(),
+                epoch_string,
+            ))
```
Comment (bmwill) [inline narwhal/primary/src/primary.rs:297]: For debugging purposes maybe we could add a SetResponseHeaderLayer setting the epoch header on all responses that narwhal returns as well?

File: narwhal/network/src/epoch_filter.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (laura-makdah) [reply narwhal/network/src/epoch_filter.rs:34]: updated

File: narwhal/network/src/epoch_filter.rs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+        let epoch = request.headers().get(EPOCH_HEADER_KEY).ok_or_else(|| {
+            Status::new_with_message(StatusCode::BadRequest, "missing epoch header").into_response()
+        })?;
+
+        if self.allowed_epoch == *epoch {
+            Ok(())
+        } else {
+            Err(
+                Status::new_with_message(StatusCode::BadRequest, "request from different epoch")
```
Comment (akichidis) [inline narwhal/network/src/epoch_filter.rs:37]: I would recommend adding on the message the `expected` epoch and `received` epoch , that will probably help us diagnose issues when looking in logs

File: narwhal/primary/src/primary.rs (hunk 1 above)
Comment (akichidis) [reply narwhal/primary/src/primary.rs:282]: So is this still needed here or was that meant to be a `SetResponseHeaderLayer` to address what @bmwill suggested?

File: narwhal/worker/src/worker.rs [hunk 6]
```diff
@@ -177,6 +188,10 @@ impl Worker {
                 inbound_network_metrics,
             )))
             .layer(CallbackLayer::new(FailpointsMakeCallbackHandler::new()))
+            .layer(SetRequestHeaderLayer::overriding(
```
Comment (akichidis) [inline narwhal/worker/src/worker.rs:191]: Similar question as above, is this needed here?

File: narwhal/network/src/epoch_filter.rs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use anemo::{types::response::StatusCode, Request, Response};
+    use anemo_tower::auth::RequireAuthorizationLayer;
+    use bytes::Bytes;
+    use tower::{BoxError, Service, ServiceBuilder, ServiceExt};
+
+    #[tokio::test]
+    async fn authorize_request_by_peer_id() {
```
Comment (akichidis) [inline narwhal/network/src/epoch_filter.rs:53]: nit: `authorize_request_by_epoch` ?

**Review by akichidis — COMMENTED** [review]
Thanks @laura-makdah , left a few comments. Is there a way to test a scenario where nodes of different epochs can't communicate?

File: narwhal/worker/src/worker.rs (hunk 6 above)
Comment (laura-makdah) [reply narwhal/worker/src/worker.rs:191]: This was recommended by Brandon for debugging

**Comment by laura-makdah** [comment]
@akichidis there is a test case for that one in epoch_filters tests

File: narwhal/network/src/epoch_filter.rs (hunk 7 above)
Comment (laura-makdah) [reply narwhal/network/src/epoch_filter.rs]: fixed

File: narwhal/network/src/epoch_filter.rs (hunk 5 above)
Comment (laura-makdah) [reply narwhal/network/src/epoch_filter.rs]: updated

**Pull request merged**